### PR TITLE
[SE-0221] Character properties

### PR DESCRIPTION
**Pull request by milseman**
https://forums.swift.org/t/pitch-character-and-string-properties/11620/68

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

**Comment by milseman** [comment]
I just measured locally, this is ~1% to ~1.5% size increase of stdlib for both character properties and scalar properties together.

**Comment by milseman** [comment]
@swift-ci please test source compatibility

**Comment by benrimmington** [comment]
Some documentation (`isNewline`, `isNumber`, `wholeNumberValue`, `isHexDigit`) uses a mixture of `///` and `//` comments.

Should the `Unicode.GeneralCategory` extension be moved to the [UnicodeScalarProperties.swift](https://github.com/apple/swift/blob/master/stdlib/public/core/UnicodeScalarProperties.swift) file?

Should `isNewline` also recognize `U+000B: LINE TABULATION (VT)` and `U+000C: FORM FEED (FF)` characters? Is there an ICU property, instead of hard-coding these?

Should `isEmoji` be available for [Apple platforms only](https://github.com/apple/swift/blob/cef828460759f233903bab0590bb74e63ccc51ef/stdlib/public/core/UnicodeScalarProperties.swift#L585-L589)? But if the emoji properties were introduced in [ICU 57](http://site.icu-project.org/download/57), they won't be available for all [supported target platforms](https://swift.org/download/#supported-target-platforms) (e.g. macOS 10.9 and iOS 7).

Should the `_crlf` and `_lf` be *stored* (rather than computed) properties?

A couple more of the boolean properties could be based on `Unicode.Scalar` APIs.

```diff
 public var isASCII: Bool {
-  return asciiValue != nil
+  return (_isSingleScalar && _firstScalar.isASCII) || (self == ._crlf)
 }
 
 public var isHexDigit: Bool {
-  return hexDigitValue != nil
+  return (_isSingleScalar && _firstScalar.properties.isHexDigit)
 }
```

**Comment by benrimmington** [comment]
For the following APIs, it might be possible to avoid temporary string creation, at least for single-scalar characters.

```diff
-public var isUppercase: Bool { return _isUppercased && isCased }
+public var isUppercase: Bool {
+  return (_isSingleScalar && _firstScalar.properties.isUppercase)
+      || (_isUppercased && isCased)
+}
 
-public var isLowercase: Bool { return _isLowercased && isCased }
+public var isLowercase: Bool {
+  return (_isSingleScalar && _firstScalar.properties.isLowercase)
+      || (_isLowercased && isCased)
+}
 
 public var isCased: Bool {
-  return !_isUppercased || !_isLowercased
+  return (_isSingleScalar && _firstScalar.properties.isCased)
+      || (!_isUppercased || !_isLowercased)
 }
```

Would it be possible to validate `isEmoji` with the contents of [emoji-test.txt](https://unicode.org/Public/emoji/latest/emoji-test.txt), if the validation-test can require the latest macOS?

**Comment by milseman** [comment]
> Some documentation (isNewline, isNumber, wholeNumberValue, isHexDigit) uses a mixture of /// and // comments.

Triple slash identifies the comment is for processing as formal documentation. As far as the individual blank line being double or triple, I'm indifferent. @natecook1000 would know more.

> Should the Unicode.GeneralCategory extension be moved to the UnicodeScalarProperties.swift file?

This is just a (functional) prototype implementation, from before those properties landed. Putting it here minimized merge conflicts, but yes, it could go there as well. I was going to ask @allevato if he was interested in the higher-level categorization.

> Should isNewline also recognize U+000B: LINE TABULATION (VT) and U+000C: FORM FEED (FF) characters? Is there an ICU property, instead of hard-coding these?

There is not, at least for general purpose in Unicode. Much of the specification and behavior surround handling them as they emerge with compatibility with other conventions.

To quote some sections: "Newlines are represented on different platforms by carriage return (CR), line feed (LF), CRLF, or next line (NEL)" ... "The Unicode Standard defines two unambiguous separator characters: U+2029 PARA- GRAPH SEPARATOR (PS) and U+2028 LINE SEPARATOR (LS). In Unicode text, the PS and LS characters should be used wherever the desired function is unambiguous." ... "Note that even if an implementer knows which characters represent NLF on a particular platform, CR, LF, CRLF, and NEL should be treated the same on input and in interpreta- tion. Only on output is it necessary to distinguish between them.".

Now, FF might be worth adding: "R4: A readline function should stop at NLF, LS, FF, or PS. In the typical implemen- tation, it does not include the NLF, LS, PS, or FF that caused it to stop."

For VT, I don't really see anything except compatibility with Microsoft Word internal document conventions. Do you see a use for including it?

> Should isEmoji be available for Apple platforms only? But if the emoji properties were introduced in ICU 57, they won't be available for all supported target platforms (e.g. macOS 10.9 and iOS 7).

At the very least, it may need some availability restrictions before merging.

> Should the _crlf and _lf be stored (rather than computed) properties?

Err, why?

> A couple more of the boolean properties could be based on Unicode.Scalar APIs.

I don't understand, what is the advantage of writing it this way?

> For the following APIs, it might be possible to avoid temporary string creation, at least for single-scalar characters.

Yes, we can write more single-scalar fast-paths. BTW, with unicode-rich small string support in the future, these will all be stored directly.

> Would it be possible to validate isEmoji with the contents of emoji-test.txt, if the validation-test can require the latest macOS?

isEmoji is careful not to perform emoji validation, but rather check for wellformedness. These are two different concerns, the former is recommended for any facility that is producing emoji rather than querying them.

**Comment by benrimmington** [comment]
> Triple slash identifies the comment is for processing as formal documentation. As far as the individual blank line being double or triple, I'm indifferent.

My concern was if the `//` makes the documentation parser ignore any portion of the comments.

> For VT, I don't really see anything except compatibility with Microsoft Word internal document conventions. Do you see a use for including it?

[UAX #14](https://www.unicode.org/reports/tr14/#BK) includes it in the BK class, but "implementations are not required to support the VT character."

It doesn't seem like an improvement, but you might be able to use `__swift_stdlib_u_getIntPropertyValue(_firstScalar.value, __swift_stdlib_UCHAR_LINE_BREAK)` and check for `U_LB_MANDATORY_BREAK`, `U_LB_CARRIAGE_RETURN`, `U_LB_LINE_FEED`, `U_LB_NEXT_LINE`, etc.

> > Should the _crlf and _lf be stored (rather than computed) properties?
>
> Err, why?

There used to be a performance issue with character literals, but it has probably been fixed by #6850 and/or #9225.

> > A couple more of the boolean properties could be based on Unicode.Scalar APIs.
>
> I don't understand, what is the advantage of writing it this way?

For example, if `_firstScalar.properties.isHexDigit` is faster than your `hexDigitValue` property.

**Comment by milseman** [comment]
> For example, if _firstScalar.properties.isHexDigit is faster than your hexDigitValue property.

The given implementation is ~6% faster, even without making hexDigitValue `@inlinable`.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test compiler performance

**Comment by benrimmington** [comment]
@milseman Can you hide my outdated comments, or shall I delete them?

(I should have used proper review comments instead).

**Comment by milseman** [comment]
@benrimmington your comments are valuable, so I'd hate to see them go away. If you must, you can put under a disclosure triangle, but I don't see any harm in leaving them here.

**Comment by benrimmington** [comment]
@milseman I don't want other reviewers to waste any time reading my comments. Someone with write permissions could minimize them: see [*Hiding a comment*](https://help.github.com/articles/managing-disruptive-comments/#hiding-a-comment). Otherwise, I'll edit or delete them.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@airspeedswift ready whenever you are

File: stdlib/public/core/CharacterProperties.swift [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+  @inlinable
+  internal var _isSingleScalar: Bool {
+    return self.unicodeScalars.count == 1
+  }
+
+  @inlinable
+  static internal var _crlf: Character { return "\r\n" }
+
+  @inlinable
+  static internal var _lf: Character { return "\n" }
```
Comment (benrimmington) [inline stdlib/public/core/CharacterProperties.swift]: The `_lf` property isn't used in this pull request.

The `_crlf` property is only used once.

File: test/stdlib/CharacterProperties.swift [hunk 2]
```diff
@@ -0,0 +1,191 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import StdlibUnittest
+
+var CharacterPropertiesTests = TestSuite("StringTests")
+
+CharacterPropertiesTests.test("ASCII queries") {
+  for cu in (0 as UInt32)...(0x7F as UInt32) {
```
Comment (benrimmington) [inline test/stdlib/CharacterProperties.swift]: Should `"\r\n"` and some non-ASCII characters also be tested?

File: test/stdlib/CharacterProperties.swift [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  expectFalse(eAccent.isUppercase)
+  expectTrue(EAccent.isUppercase && EAccent.isCased)
+  expectFalse(EAccent.isLowercase)
+
+  expectTrue(Character("И").isUppercase)
+  expectTrue(Character("и").isLowercase)
+  expectTrue(Character("Π").isUppercase)
+  expectTrue(Character("π").isLowercase)
+
+  expectEqual("SS", Character("ß").uppercased())
```
Comment (benrimmington) [inline test/stdlib/CharacterProperties.swift]: Is the [Capital ẞ](https://en.wikipedia.org/wiki/Capital_ẞ) (U+1E9E: LATIN CAPITAL LETTER SHARP S) supported by ICU yet?

```
expectEqual("ß", Character("ẞ").lowercased())
```

The system font (SF Pro) displays it as an "SS" ligature.

**Review by airspeedswift — APPROVED** [review]
I think we probably need benchmarks too (though they can land separately). Particularly of the numeric conversion parts.

**Comment by milseman** [comment]
Err, not sure why I closed this PR, but it can't be re-opened now that there was an update. Tracking in https://github.com/apple/swift/pull/20520

**Pull request closed without merging**